core/services/ocr2/plugins/ocr2vrf/blockhashes/fixed_blockhash_provider.go

```diff
@@ -41,59 +49,77 @@ func (b *fixedBlockhashProvider) OnchainVerifiableBlocks(
 	if err != nil {
 		return 0, nil, errors.Wrap(err, "current height")
 	}
-	fromBlock := toBlock - b.lookbackBlocks
-	var reqs []rpc.BatchElem
 
+	fromBlock := uint64(0)
+	if toBlock > b.lookbackBlocks {
+		fromBlock = toBlock - b.lookbackBlocks
+	}
+
+	var blockHeights []uint64
 	for i := fromBlock; i <= toBlock; i++ {
-		req := rpc.BatchElem{
-			Method: "eth_getBlockByNumber",
-			Args:   []interface{}{hexutil.EncodeBig(big.NewInt(int64(i))), false},
-			Result: &evmtypes.Head{},
+		blockHeights = append(blockHeights, i)
+	}
+
+	blockhashes := make([]common.Hash, len(blockHeights))
+
+	heads, err := b.lp.GetBlocks(blockHeights, pg.WithParentCtx(ctx))
+	if err != nil {
+		b.logger.Warnw("error in get blocks", "err", err)
+	}
+
+	for _, h := range heads {
+		blockhashes[h.BlockNumber-int64(fromBlock)] = h.BlockHash
+	}
+
+	// Fallback to RPC for blocks not found in log poller
+	var reqs []rpc.BatchElem
+	for i, bh := range blockhashes {
+		if bh == utils.EmptyHash {
+			req := rpc.BatchElem{
+				Method: "eth_getBlockByNumber",
+				Args:   []interface{}{hexutil.EncodeBig(big.NewInt(int64(i) + int64(fromBlock))), false},
+				Result: &evmtypes.Head{},
+			}
+			reqs = append(reqs, req)
 		}
-		reqs = append(reqs, req)
 	}
 
 	for i := 0; i < len(reqs); i += int(b.batchSize) {
 		j := i + int(b.batchSize)
 		if j > len(reqs) {
 			j = len(reqs)
 		}
-
 		err := b.client.BatchCallContext(ctx, reqs[i:j])
-
 		if err != nil {
 			return 0, nil, errors.Wrap(err, "batch call context eth_getBlockByNumber")
 		}
 	}
 
-	var blockhashes []common.Hash
 	for _, r := range reqs {
 		if r.Error != nil {
 			return 0, nil, errors.Wrap(r.Error, "error found in eth_getBlockByNumber response")
 		}
 		block, is := r.Result.(*evmtypes.Head)
+
 		if !is {
 			return 0, nil, errors.Errorf("expected result to be a %T, got %T", &evmtypes.Head{}, r.Result)
 		}
 		if block == nil {
 			return 0, nil, errors.New("invariant violation: got nil block")
 		}
-		if block.Hash == (common.Hash{}) {
+		if block.Hash == utils.EmptyHash {
 			return 0, nil, errors.New("missing block hash")
```

Comment:
Can this error have more context? e.g `fmt.Errorf("missing block hash for block %d", hexutil.DecodeBig(r.Args[0]))`? I think this should be okay, but not 100% sure.

Reply:
yeah I'd be curious to see what the value is in this exceptional case. Will make the change

Reply:
I wouldn't put it past RPC's to hit this exceptional case 😉 .